proposals/0000-session-data-extension.md

```diff
@@ -0,0 +1,63 @@
+
+Title: **DEP-0000: Session Data (Extension Message)**
+
+Short Name: `0000-session-data-extension`
+
+Type: Informative
+
+Status: Undefined (as of 2018-05-31)
+
+Github PR: https://github.com/datprotocol/DEPs/pull/27
+
+Authors: [Paul Frazee](https://github.com/pfrazee)
+
+
+# Summary
+[summary]: #summary
+
+This DEP defines the non-standard `session-data` extension message used in the Dat replication protocol. This message provides a way to attach application data to a connection, commonly used for identifying the users and broadcasting personal keys. 
+
+
+# Motivation
+[motivation]: #motivation
+
+Applications frequently need to discover which users of the application are online (presence) in order to establish bidirectional communication. For example, a chat application which uses a shared HyperDB as the channel state may need to broadcast the Hypercore keys of each user in order to authorize the joining chat-users (as in the case of "Cabal"). It would also be useful to broadcast Hyperdrive archive keys (as in the case of "Fritter" and "Rotonde") or even simple plain-text identity (eg "my name is Bob") to be used with other communication mechanisms.
+
+This extension message will establish a common mechanism for broadcasting user and session data.
+
+
+# Reference Documentation
+[reference-documentation]: #reference-documentation
+
+This DEP is implemented using the Dat replication protocol's "extension messages." In order to broadcast support for this DEP, a client should declare the `'session-data'` extension in the replication handshake.
+
+Session-data can be announced at any time after the connection is established by sending an extension message of type `'session-data'`. The message may include a payload up to 256 bytes in length. Any additional bytes should be truncated by the receiving client. The payload is a buffer of any encoding. The session-data message should not be sent frequently and a client may choose to rate-limit its handling of the events (this DEP suggests "once per 5 seconds").
+
+The client should maintain a `sessionData` variable on each connection. This variable should be empty when a new connection is established. Any time a `'session-data'` extension message is received, the value of the `sessionData` variable should be updated to contain the payload of the message.
+
+The client may respond to the message by emitting an event, so that it may be handled by the client's application logic. The client should also make the most recent `sessionData` buffer available to the application logic after message is received.
+
+After publishing this DEP, the "Beaker Browser" will implement a Web API for exposing the `'session-data'` protocol to applications. It will restrict access so that the application code of a `dat://` site will only be able to set the session data for connections related to its own content.
```

Comment:
What are "connections related to it's own content"? Is it the DatArchive for the URL the content is being served from?

I think that any DatArchives that are created should support session data to support use cases like chat where people might have different apps (in different dats) all talking via the same protocols over one URL.

Reply:
> Is it the DatArchive for the URL the content is being served from?

Correct

> I think that any DatArchives that are created should support session data to support use cases like chat where people might have different apps (in different dats) all talking via the same protocols over one URL.

That makes sense; you're right that this is an issue. I need to think about the security implications. It's also not clear to me how that would work: how would an app know which other dat archives it needs to be looking at to get the session data?

I think solving that would require some meta-identifier which the sessions are being attached to, where you say "this is `fritter` session data, anybody else interested in `fritter` should receive this." And then we'd need to know which connections are interested in `fritter` session data so that we know which connections should receive the message. You also create the possibility for multiple apps to register session data. It gets a much more complex.

Reply:
Hmm, in fact if you have 2 versions of the fritter app, the only common connections would be for the profile archives.

Reply:
That's a good point. I could see it getting hairy pretty quick. The ephemeral messages spec wouldn't have the problem of specifying where the data is coming from since that can be done at the application level.

> how would an app know which other dat archives it needs to be looking at to get the session data

I think the UX would be copy-pasting the link for the channel you want to use for communication. Then it would open the dat archive, which would probably have info about the channel name or whatever, and start looking for peers to talk to (or do the multiwriter thing from cabal),

Alternatively when people make a chat channel, they can fork the application Dat and send the link to that instead of having the application dat. Though that would make updates to the content harder, and would mean having multiple windows for chatting instead of a single one. (Might be fixable with iframes?)

I guess in the worst case, the main application logic for getting the data can live in an iframe and talk to the parent responsible for the UI.

Overall it's not a dealbreaker, but it'll require being a little more "clever" in the implementation.

Reply:
Some discussion from IRC. TL;DR: I'm starting to think this may not work and ephemeral messaging may need to be the solution. The problem is, how do you handle multiple apps?

<details>

```
pfrazee> in the ephemeral messaging case, you're going to have
apps cross-talking with each other. Imagine two apps sending messages
on the same channel and receiving responses intended for one or the
other

pfrazee> that'd be very confusing

RangerMauve> I think multiple apps, or even multiple pages of the same
app, writing to the same persistent variable will cause problems

RangerMauve> pfrazee: Maybe have a "stream ID" to have multiplexing be
a first-class concern?

pfrazee> yeah exact same issue, though multiple pages of the same
app *should* be able to coordinate enough not to contest with each
other (because it OUGHT to be setting the same thing)

pfrazee> perhaps

pfrazee> the complexity increase of all the possible solutions
concerns me

RangerMauve> I guess the main issue here is that beaker has a single
replication stream for all pages.

RangerMauve> For lower-level one-app-to-one-stream use cases what you
sketched out works just fine

pfrazee> yeah

pfrazee> so on fritter, for instance, I'd want to attach my
profile key to the dat channel for each profile dat I sync

RangerMauve> pfrazee: To let them know you're following them?

pfrazee> yeah I figure that's how I'd do that

pfrazee> and then the app would probably read the keys attached to
their personal dat archive

RangerMauve> Yeah, that's pretty elegant

pfrazee> yeah then you just gotta figure out how you deal with
multiple apps

pfrazee> I almost wonder if the solution is to let multiple
sessions be attached

RangerMauve> I think ephemeral messaging with application-level RPC
would be a good approach. If you get a message, and it doesn't make
sense to you, ignore it.

pfrazee> well the downside there is, ephemeral messages can easily
be missed because you may not have the app open when it's sent

RangerMauve> Yeah. But that can be accounted for at the application
level, too.

RangerMauve> For example, ping all contact dats when you open the
page, and react to pings while you're active

pfrazee> yeah that's probably how you'd have to do it

RangerMauve> If it's already considerd an "unreliable" channel, then
applications will already need some sort of mechanisms in place to
account for "missed" messages

RangerMauve> And "duplicated" messages

pfrazee> well that's what session-data semantics helps fix

pfrazee> persists the data and it's atomic so you can resend

RangerMauve> I think that for reliable transports, they can create a
throwaway dat and post messages in there, then forget about it when
it's no longer relevant

RangerMauve> Yeah, but that would only work for cases where one
application is writing to the session.

pfrazee> I guess you can "multiplex" the sessions

pfrazee> allow multiple sessionData's to be attached and then
beaker would just let each origin write only one

pfrazee> but that does start to feel...weird

RangerMauve> Yeah, I'm not sure how clean that would be in the end

pfrazee> yeah. I'll talk to mafintosh about this tonight. You are
starting to make me think we'll have to do ephemeral messages instead
and let apps solve it
```

</details>

Reply:
Oh, more ideas:

What if each message required a 20 byte key for the "type" of message it was, similar to how discovery-swarm uses keys for discovery and truncates to 20 characters.

Then, an application would be required to pass the key it wants to communicate on to automagically filter incoming messages.

Multiple instances using the same key will still need to figure out deduplication in the protocol, but it would require an application to explicitly opt-into a type of message.

Reply:
Regarding security:

- If an application already knows the dat URL it wants to talk on, then it's probably allowed to talk on that URL.
- You could require user action for allowing an application to listen on a type of message.

Reply:
I'm wondering if we should stick with the Single Origin Policy in the DEP (can only attach session-data to connections for your own dat), write another DEP for ephemeral messaging too with the same S.O.P. approach, and then just expect that we'll need a more sophisticated solution to support apps talking to each other.

My goal is really to come up with something that's simple and works *well enough* for simple apps to "self communicate" among peers, so to speak. I'm not yet sure whether I think this will be a long-term solution for discovery, so I'm not sure we should spend time agonizing over other use-cases. We can always supersede this DEP.

Reply:
Yeah, working around SOP won't be too hard with iframes.

Reply:
@RangerMauve (I believe) that's only possible if the target Dat has JS which helps you do it.

Reply:
Yeah, which is a good thing because then whoever made that dat will need to explicitly opt-into the functionality. I'm making use of [frame-rpc](https://github.com/substack/frame-rpc) for this use case in dat-polyfill, actually.